pennylane/gradients/parameter_shift_cv.py

```diff
@@ -0,0 +1,678 @@
+    obs_indices = []
+    gradient_values = []
+
+    for idx, _ in enumerate(tape.trainable_params):
+        t_idx = list(tape.trainable_params)[idx]
+        op = tape._par_info[t_idx]["op"]
+
+        if idx not in argnum:
+            # parameter has zero gradient
+            shapes.append(0)
+            obs_indices.append([])
+            gradient_values.append([])
+            continue
+
+        shapes.append(1)
+
+        # get the gradient recipe for the trainable parameter
+        recipe = gradient_recipes[argnum.index(idx)]
+        recipe = recipe or _get_operation_recipe(tape, idx, shift=shift)
+        recipe = _process_gradient_recipe(recipe)
+        coeffs, multipliers, shifts = recipe
+
+        if len(shifts) != 2:
+            # The 2nd order CV parameter-shift rule only accepts two-term shifts
+            raise NotImplementedError(
+                "Taking the analytic gradient for order-2 operators is "
+                f"unsupported for operation {op} which has a "
+                "gradient recipe of more than two terms."
+            )
+
+        shifted_tapes = generate_shifted_tapes(tape, idx, shifts, multipliers)
+
+        # evaluate transformed observables at the original parameter point
+        # first build the Heisenberg picture transformation matrix Z
+        Z0 = op.heisenberg_tr(dev_wires, inverse=True)
+        Z2 = shifted_tapes[0]._par_info[t_idx]["op"].heisenberg_tr(dev_wires)
+        Z1 = shifted_tapes[1]._par_info[t_idx]["op"].heisenberg_tr(dev_wires)
+
+        # derivative of the operation
+        Z = Z2 * coeffs[0] + Z1 * coeffs[1]
+        Z = Z @ Z0
+
+        # conjugate Z with all the descendant operations
+        B = np.eye(1 + 2 * len(dev_wires))
+        B_inv = B.copy()
+
+        succ = tape.graph.descendants_in_order((op,))
+        operation_descendents = itertools.filterfalse(qml.circuit_graph._is_observable, succ)
+        observable_descendents = filter(qml.circuit_graph._is_observable, succ)
+
+        for BB in operation_descendents:
+            if not BB.supports_heisenberg:
+                # if the descendant gate is non-Gaussian in parameter-shift differentiation
+                # mode, then there must be no observable following it.
+                continue
+
+            B = BB.heisenberg_tr(dev_wires) @ B
+            B_inv = B_inv @ BB.heisenberg_tr(dev_wires, inverse=True)
+
+        Z = B @ Z @ B_inv  # conjugation
+
+        g_tape = tape.copy(copy_operations=True)
+        constants = []
+
+        # transform the descendant observables into their derivatives using Z
+        transformed_obs_idx = []
+
+        for obs in observable_descendents:
+            # get the index of the descendent observable
+            idx = tape.observables.index(obs)
+            transformed_obs_idx.append(idx)
+
+            transformed_obs = _transform_observable(obs, Z, dev_wires)
+
+            A = transformed_obs.parameters[0]
+            constants.append(None)

```

Comment:
(Minor): how about appending `None` only when there is no constant?

If I understand it correctly, first we append `None` and then remove it in cases when there is a constant term. This logic seems to be slightly more prone to confusion.

Reply:
I originally did that, but then it causes issues further down as you are unable to use the `constants` list in the `zip` 🤔 

So it seems to be a tradeoff; make it simpler here at the expense of more complicated `zip` logic later, or make it longer here so that it works in the `zip`.

Reply:
Oh really? I would have expected that modifying the if structure to always append to `constants` and extending it with two `else` branches could help :thinking: 
```python
            # Check if the transformed observable corresponds to a constant term.
            if len(A.nonzero()[0]) == 1:
                if A.ndim == 2 and A[0, 0] != 0:
                    constants.append(A[0, 0])

                elif A.ndim == 1 and A[0] != 0:
                    constants.append(A[0])
                else:
                    constants.append(None)
            else:
                constants.append(None)
```

Reply:
Ah, I see what you mean @antalszava!

I wonder if it is slightly slower though, due to the cost of the additional branching? Not sure, and if it is, it is probably very small 😆

Reply:
> I wonder if it is slightly slower though, due to the cost of the additional branching?

Good question! Maybe since these are `else`s, those are less costly. With the current version, we might do some extra list operations. Not sure about the overall performance implications. :thinking:

Reply:
I updated it to this, which I think is cleaner :)

![image](https://user-images.githubusercontent.com/2959003/128297139-53c8ee1e-4443-4812-b5b8-71c875426d8e.png)